crates/deltalake-core/src/operations/convert_to_delta.rs

```diff
@@ -0,0 +1,807 @@
+        // Return an error if neither object store nor path is provided
+        let object_store = if let Some(object_store) = self.object_store {
+            object_store
+        } else if let Some(location) = self.location {
+            Arc::new(DeltaObjectStore::try_new(
+                ensure_table_uri(location)?,
+                self.storage_options.unwrap_or_default(),
+            )?)
+        } else {
+            return Err(Error::MissingLocation);
+        };
+
+        // Return an error if the location is already a Delta table location
+        if object_store.is_delta_table_location().await? {
+            return Err(Error::DeltaTableAlready);
+        }
+        info!(
+            "Converting Parquet table in object store location: {:?}",
+            object_store.root_uri()
+        );
+
+        // Get all the parquet files in the location
+        let mut files = Vec::new();
+        let fut = object_store
+            .list(None)
+            .await?
+            .try_for_each_concurrent(10, |meta| {
+                if Some("parquet") == meta.location.extension() {
+                    debug!("Found parquet file {:#?}", meta.location);
+                    files.push(meta);
+                }
+                future::ready(Ok(()))
+            });
+        // Return immediately with an error if the stream returns an error
+        fut.await?;
+        if files.is_empty() {
+            return Err(Error::ParquetFileNotFound);
+        }
+
+        // Iterate over the parquet files. Parse partition columns, generate add actions and collect parquet file schemas
+        let mut arrow_schemas = Vec::new();
+        let mut actions = Vec::new();
+        // A HashSet of all unique partition columns in a Parquet table
+        let mut partition_columns = HashSet::new();
+        // A vector of SchemaField of all unique partition columns in a Parquet table
+        let mut partition_schema_fields = Vec::new();
+        for file in files {
+            // A HashMap from partition column to value for this parquet file only
+            let mut partition_values = HashMap::new();
+            let mut iter = file.location.as_ref().split('/').peekable();
+            let mut subpath = iter.next();
+            // Get partitions from subpaths. Skip the last subpath
+            while iter.peek().is_some() {
+                if let Some(subpath) = subpath {
+                    let partition = DeltaTablePartition::try_from(
+                        percent_decode_str(subpath).decode_utf8()?.as_ref(),
+                    )?;
+                    debug!(
+                        "Found partition {partition:#?} in parquet file {:#?}",
+                        file.location
+                    );
+                    let (key, val) = (partition.key, partition.value);
+                    partition_values.insert(
+                        key.clone(),
+                        if val == NULL_PARTITION_VALUE_DATA_PATH {
```

Comment:
Haven't dealt with many alternative partitioning schemes, but I guess this would limit `ConvertToDelta` to only use hive-partitioned tables. Not sure about the implications, but maybe we could make the partitioning scheme configurable, but have only hive partitioning as an option for now?

Reply:
Good idea. I added a `PartitionStrategy` enum so that the partitioning scheme is configurable. For now, the enum only has a `Hive` variant. It appears that [`ConvertToDeltaCommand.scala`](https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/commands/ConvertToDeltaCommand.scala#L319) also only supports hive-partitioning for parquet paths. If a parquet path is non-hive, `ConvertToDeltaCommand.scala` will throw an error. This aligns with [`DeltaTablePartition::try_from`](https://github.com/delta-io/delta-rs/pull/1686/files#diff-55016b8b61e61c4b987804307f47377cda9bde37a3dcb0e014ce4f27710bd212R252). 
I created a follow-up [issue](https://github.com/delta-io/delta-rs/issues/1828) to extend `PartitionStrategy` to support other popular partitioning schemes such as the directory scheme like `./partitioned/2000/part-0.parquet`. User will need to provide the column name when they use this scheme.